data_source.go

```diff
@@ -303,38 +303,56 @@ func (ds *pgDataSource) buildConnInfo(conninfo map[string]string) string {
 		ds.cfg.Database, ds.cfg.User, ds.cfg.Sslmode, ds.cfg.BinaryParameters, ds.cfg.Password)
 }
 
-func (ds *pgDataSource) getPgCurrentWalLsn(role string) (string, error) {
-	if role == "replica" {
-		// query select * from pg_stat_wal_receiver;  to get conninfo
-		//create a new db connection to upstream (primary) with conninfo and return pg_current_wal_lsn
-		conninfo, err := ds.getUpstreamConnInfo()
-		if err != nil {
-			return "", err
-		}
-		upstreamConnInfo := ds.buildConnInfo(parseConnInfo(conninfo))
-		upstreamDb, err := sqlConnect(upstreamConnInfo)
-		if err != nil {
-			return "", err
-		}
-		defer upstreamDb.Close()
+func (ds *pgDataSource) getPgCurrentWalLsn(role string, maxHop int64) (string, error) {
+	db, dbErr := ds.getDB()
+	if dbErr != nil {
+		return "", dbErr
+	}
+	if maxHop == 0 {
 		var pgCurrentWalLsn string
-		err = upstreamDb.Get(&pgCurrentWalLsn, "select pg_current_wal_lsn()")
+		err := db.Get(&pgCurrentWalLsn, "select pg_current_wal_lsn()")
```

Comment:
If we use maxHop == 0 as the base case, does that default max hop is 2? 

This db is the local db, which, if it's a replica, will fail to execute current wal lsn.

Reply:
ah great points! I will change the base case to `maxHop == 1`.

Also, I will update the base case to check whether its a replica or not.

Reply:
@film42 should be better now